### PR TITLE
search: add basic parameter scanning to query parser

### PR DESCRIPTION
**Pull request by rvantonder**
This PR changes the previous

```
type Parameter {
	Value string
}
```
to
```
type Parameter struct {
	Field   string
	Value   string
	Negated bool
}
```

I thought deeply about where to draw the line of parsing a "parameter" at this low level parsing stage. Please see comment for `ScanParameter` for what it entails. I chose this way because I'd like our literal parser (or _any_ parser, ever, really) to have a low level representation of a 'parameter' that can then decide what it considers to be valid or not. The logic is, in essence, a minimal unvalidated interpretation that doesn't lose or throw away (e.g. by, unquoting, etc.) any syntactic information that could potentially be useful in a search, and cannot fail. For example, if literal search wanted to search for `"`, it should be able to get that value in a leaf node by using `ScanParameter`. And, indeed, the current approach allows this.

I fully plan to migrate our literal search parser to use `ScanParameter`. Note that `ParseParameter` is the caller that is specific to AND/OR query parsing (it is less general than `ScanParameter`) which is why these two functions are kept separate.

---

A note on `NOT`: don't think about it in the context of this PR right now. When introduced, it will serve as syntactic sugar, specific to AND/OR queries, that influences the `Negated` field of a parameter. In the future, `NOT` may apply to whole expressions, which will then be promoted to it's own `Operator`, rather than syntactic sugar on `Parameter`s

### CODE REVIEW

File: internal/search/parser.go [hunk 1]
```diff
@@ -184,18 +227,16 @@ func (p *parser) parseParameterList() ([]Node, error) {
...
-			value, err := p.scanParameter()
-			if err != nil {
-				return nil, err
-			}
-			nodes = append(nodes, Parameter{Value: value})
+			parameter := p.ParseParameter()
+			nodes = append(nodes, parameter)
 		}
 	}
+Done:
```
Comment (rvantonder) [inline internal/search/parser.go]: Dijkstra would be proud. But seriously: I just want to break out of the loop, and `break` won't do that.

File: internal/search/parser.go [hunk 2]
```diff
@@ -142,10 +149,46 @@ func (p *parser) skipSpaces() error {
...
+// When form (1) does not match, Value corresponds to <string> and Field is the empty string.
+//
+// The value parameter in the parse tree is only distinguished with respect to
+// the two forms above. There is no restriction on values that <string> may take
+// on. Notably, there is no interpretation of quoting or escaping, which may vary
+// depending on the search being performed. All validation with respect to such
+// properties, and how these should be interpretted, is thus context dependent
+// and handled appropriately within those contexts.
+func ScanParameter(parameter []byte) Parameter {
+	fieldValuePattern := regexp.MustCompile("(^-?[a-zA-Z0-9]+):(.*)")
```
Comment (rvantonder) [inline internal/search/parser.go]: I can do this in a loop over runes for the unicode friendliness. The important part is that the interface and tests of this function will remain the same. Will do in a follow up.

File: internal/search/parser.go [hunk 3]
```diff
@@ -27,7 +28,9 @@ func (Operator) node()  {}
 
 // Parameter is a leaf node of expressions.
 type Parameter struct {
-	Value string
+	Field   string `json:"field"`
+	Value   string `json:"value"`
+	Negated bool   `json:"negated"`
```
Comment (unknwon) [inline internal/search/parser.go]: Would be nice to have docstrings for these fields, or even one example could be more intuitive.

File: internal/search/parser.go [hunk 4]
```diff
@@ -44,7 +47,11 @@ type Operator struct {
 }
 
 func (node Parameter) String() string {
-	return node.Value
+	if node.Field == "" {
+		return node.Value
+	} else {
```
Comment (unknwon) [inline internal/search/parser.go]: nit: Go idiomatically, two `return`s do not require an `else`

File: internal/search/parser.go (hunk 2 above)
Comment (unknwon) [reply internal/search/parser.go]: Good place to use `lazyregexp`? See examples: https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+lazyregexp&patternType=literal

I don't think we have a hard enforcement on this, but lots of `regexp.MustCompile` slows start of the server.

File: internal/search/parser.go (hunk 1 above)
Comment (unknwon) [reply internal/search/parser.go]: You can move `Done:` to line 209 (immediate line before the `for`), then do `break Done`. In this case, maybe rename it to `loop:`, `break loop` feels more nature.

File: internal/search/parser.go (hunk 2 above)
Comment (ryanslade) [reply internal/search/parser.go]: We should also move the declaration outside of the func so that we don't recompile the pattern every time the function is called

Comment (keegancsmith) [reply internal/search/parser.go]: yeah this being compiled once is a good optimization, and lazyregexp is the way to do it.

Comment (keegancsmith) [reply internal/search/parser.go]: this should still be unicode friendly? ie the `field` can be the ascii chars you specify, then the parts after `:` can be anything so will be unicode/anything friendly.

File: internal/search/parser_test.go [hunk 5]
```diff
@@ -1,12 +1,82 @@
...
+func Test_ScanParameter(t *testing.T) {
+	cases := []struct {
+		Name  string
+		Input string
+		Want  string
+	}{
+		{
+			Name:  "Normal field:value",
+			Input: `field:value`,
+			Want:  `{"field":"field","value":"value","negated":false}`,
```
Comment (keegancsmith) [inline internal/search/parser_test.go]: reading this json is hard. Why don't you make want a Parameter? `cmd.Diff` will also produce nicer diffs in case of failure.

If you are worried about multiple lines and readability, you can create some helper functions to have a mini dsl :)

```go
pair := func(field, value string) Parameter { ... }
negated_pair := func(field, value string) Parameter { ... }
value := func(value string) Parameter { ... }
```

File: internal/search/parser_test.go [hunk 6]
```diff
@@ -1,12 +1,82 @@
...
 
 	"github.com/google/go-cmp/cmp"
 )
 
+func Test_ScanParameter(t *testing.T) {
+	cases := []struct {
+		Name  string
+		Input string
+		Want  string
+	}{
```
Comment (keegancsmith) [inline internal/search/parser_test.go]: nit: for a function like this I'd suggest not naming each test case, and instead just doing `cases := map[string]Parameter{...}`. The input to the parameter is enough to understand what is being tested.

File: internal/search/parser.go [hunk 7]
```diff
@@ -142,10 +149,46 @@ func (p *parser) skipSpaces() error {
...
-// scanParameter scans for leaf node values.
-func (p *parser) scanParameter() (string, error) {
+// ScanParameter returns a leaf node value usable by _any_ kind of search (e.g.,
+// literal or regexp, or...) and always succeeds.
+//
+// A parameter is a contiguous sequence characters, where the following two forms are distinguished:
+// (1) a string of syntax field:<string> where : matches the first encountered colon, and field must match ^-?[a-zA-Z0-9]+
+// (2) <string>
+//
+// When a parameter is of form (1), the string corresponds to Field:Value in type Parameter, and negated if Field starts with '-'.
```
Comment (mrnugget) [inline internal/search/parser.go]:
```suggestion
// When a parameter is of form (1), the <string> corresponds to Parameter.Value, field corresponds to Parameter.Field and Parameter.Negated is set if Field starts with '-'.
```

File: internal/search/parser.go [hunk 8]
```diff
@@ -142,10 +149,46 @@ func (p *parser) skipSpaces() error {
...
+				Negated: true,
+			}
+		}
+		return Parameter{Field: string(result[1]), Value: string(result[2])}
+	}
+	return Parameter{Field: "", Value: string(parameter)}
+}
+
+// ParseParameter returns valid leaf node values for AND/OR queries, taking into
+// account escape sequencies for special syntax: whitespace and parentheses.
```
Comment (mrnugget) [inline internal/search/parser.go]:
```suggestion
// account escape sequences for special syntax: whitespace and parentheses.
```

File: internal/search/parser_test.go [hunk 9]
```diff
@@ -1,12 +1,82 @@
...
+		{
+			Name:  "Normal field:value",
+			Input: `field:value`,
+			Want:  `{"field":"field","value":"value","negated":false}`,
+		},
```
Comment (mrnugget) [inline internal/search/parser_test.go]: Tiniest nitpick: I would use real fields and real values in the test input, so it's easier to read and understand what's parsed into what.

```suggestion
		{
			Name:  "Normal field:value",
			Input: `repohasfile:README.md`,
			Want:  `{"field":"repohasfield","value":"README.md","negated":false}`,
		},
```

File: internal/search/parser_test.go [hunk 10]
```diff
@@ -1,12 +1,82 @@
...
+		{
+			Name:  "Minus in the middle is not a valid field",
+			Input: `fie-ld:value`,
+			Want:  `{"field":"","value":"fie-ld:value","negated":false}`,
+		},
+		{
+			Name:  "No effect on whitespace",
+			Input: `  a pattern  `,
+			Want:  `{"field":"","value":"","negated":false}`,
```
Comment (mrnugget) [inline internal/search/parser_test.go]: I don't understand why `value` is `""` here. Shouldn't it be `a pattern`?

File: internal/search/parser.go [hunk 11]
```diff
@@ -142,10 +149,46 @@ func (p *parser) skipSpaces() error {
...
+func (p *parser) ParseParameter() Parameter {
 	start := p.pos
 	for {
+		if p.expect(`\ `) || p.expect(`\(`) || p.expect(`\)`) {
```
Comment (tsenart) [inline internal/search/parser.go]: Getting into parser combinator like territory, we could have something like:
```suggestion
		if p.expectAny(`\ `, `\(`, `\)`) {
```

File: internal/search/parser.go (hunk 1 above)
Comment (tsenart) [reply internal/search/parser.go]: How about just `return nodes, nil` like you had before?

File: internal/search/parser_test.go (hunk 6 above)
Comment (tsenart) [reply internal/search/parser_test.go]: I'd still keep the `[]struct`, rather than using map, for stable test execution ordering.

File: internal/search/parser.go (hunk 2 above)
Comment (rvantonder) [reply internal/search/parser.go]: I didn't use the lazy package at first because I was fuzzing this separately and didn't want to bother with pulling in other package dependencies. Restored it.

File: internal/search/parser.go (hunk 1 above)
Comment (rvantonder) [reply internal/search/parser.go]: @tsenart, you're right! I think I was doing something before that needed this, but then that issue disappeared after other changes 🎆

File: internal/search/parser_test.go (hunk 10 above)
Comment (rvantonder) [reply internal/search/parser_test.go]: Ah I forgot to remove this. This is actually a useless test because I'm calling `ParseParameter` that breaks on leading whitespace to return control to the parser. Basically, it is guaranteed that calling `ParseParameter` should not start with whitespace. If I called/tested `ScanParameter` instead it should (and would) be `  a pattern  ` (preserving the whitespace). But in this change I'm only testing `ParseParameter`. `ScanParameter` will be used with the literal parser and testing will follow there.

File: internal/search/parser_test.go (hunk 6 above)
Comment (rvantonder) [reply internal/search/parser_test.go]: I feel like it's always useful to have a test failure tell me which test case fails. The significance of the name ("what it does") is less important than having an identifier. I.e., if I called it `1` that'd be fine to. It is tiresome if I run a test and get:

```
Test failed: parser_test.go:69: 
```
which points only to `t.Error(diff)` and without comparing test inputs/outputs I don't know which case it is. I much prefer the utility of

```
--- FAIL: Test_ScanParameter (0.00s)                                                                                                                                                                               
    --- FAIL: Test_ScanParameter/Last_char_is_colon (0.00s)                                                                                                                                                        
        parser_test.go:69:   string(          
```

Where I can see which test case it was. The need for a name is less important when all tests pass, but it's particularly when tests fail (i.e., when writing them, or changing code) that these names become useful. Am I missing some other mechanism that will give me this info?

File: internal/search/parser.go (hunk 11 above)
Comment (rvantonder) [reply internal/search/parser.go]: Yep. Not going to focus on providing a better utility function for this right now, but will keep in mind.

File: internal/search/parser_test.go (hunk 5 above)
Comment (rvantonder) [reply internal/search/parser_test.go]: Hehe. Well, the story is I don't like writing test outputs. Every time I write expected outputs of a test I ask myself "Why am I doing this? Why is this needed? Such a waste of time." Because, in practice, at some point during testing you know what the oracle for a test output should be, and we can get that output, conceptually, by running the function on the input. Why go and manually enter that oracle? Annoying. So, instead, you make a test target a string representation, and, when the test fails initially (you haven't told the test what the expected value is yet) it will give you a diff string, you copy-paste the string output to the test, and you're done. I'm used to this workflow, and then also one where the tests output is copied into the test files (so there is zero manual steps). Here, I just want a string representation that I can copy over, and am OK performing the copy step manually. JSON is just a side effect to get a serialized form that is resilient to change. So: I'll change this to use the `parameter.String()` method for readability, but I like that less for the case where someone might change the `String()` representation (and would then need to update the tests), versus using JSON which wouldn't change.

Contrast with using `Parameter`: Now, I can enter a blank `Parameter` for a test case without having test output initially. I run the test, and then the diff test fails and says: 

```
        parser_test.go:70:   interface{}(                                                                                                                                                                          
            -   s"",                                                                                                                                                                                               
            +   string("field:value"),                                                                                                                                                                             
              )                                                                                                                                                                                                    
```

Well, that `string("field:value")` output isn't something I can copy over as a test oracle. So now I have to still go and manually write out `Parameter{ Field: ...` as the expected value. It's just inefficient, and when the test breaks, you also can't rely on the diff output to help you construct a new oracle.

Comment (rvantonder) [reply internal/search/parser_test.go]: Also, with the workflow above, I'd argue that the actual test representation (`Want` string value) is less significant in terms of readability, because it is so easy to reproduce the correct output. For when the test fails, it's possible to get a perfectly good looking diff for JSON formats. Since that's the main concern here, I'll use `String()` as I said, and hopefully `cmp.Diff` can produce better things with that.

Comment (rvantonder) [reply internal/search/parser_test.go]: Ah: no, I'm going back on the `String()` thing. When I use that I get the following output updates (predictably):

```
        parser_test.go:68:   string(                                                                                                                                                                               
            -   `{"field":"field","value":"value","negated":false}`,                                                                                                                                               
            +   "field:value",                                                                                                                                                                                     
              )                                                                                                                                                                                                    
                                                                                                                                                                                                                   
    --- FAIL: Test_ScanParameter/First_char_is_colon (0.00s)                                                                                                                                                       
        parser_test.go:68:   string(                                                                                                                                                                               
            -   `{"field":"","value":":value","negated":false}`,                                                                                                                                                   
            +   ":value",                                                                                                                                                                                          
              )  
...
```

which loses information. I'd like to preserve the function of `String()` to give back the actual syntactic representation so that we can use that to faithfully print out the original syntactic query. I'm open to an alternative string representation versus JSON, but don't have an immediate suggestion (and would want to call a function that gives me that something other than `String()`). This is why JSON works out to be a convenient option...

File: internal/search/parser.go (hunk 4 above)
Comment (rvantonder) [reply internal/search/parser.go]: Hehe nice. Still not used to writing if-else expressions :P

File: internal/search/parser_test.go (hunk 9 above)
Comment (rvantonder) [reply internal/search/parser_test.go]: rewrite_all_the_test_inputs.gif

Comment (rvantonder) [reply internal/search/parser_test.go]: >  I would use real fields and real values in the test input

I think part of the issue is that there aren't good real values for the things I'm testing:

`field:value:value`, `-:value`, ... but I'll try.

File: internal/search/parser_test.go (hunk 6 above)
Comment (keegancsmith) [reply internal/search/parser_test.go]: this is a nit, so not important to solve. But your `t.Error` can say the input. My point is the input to the function describes the test case perfectly. As you add more test cases, it will become increasingly difficult to think up a "descriptive name", and you will end up doing things like "Last char is colon 2". You don't really need each case to be a subtest (ie a `t.Run`. For a similar example see https://golang.org/src/regexp/syntax/parse_test.go

Comment (rvantonder) [reply internal/search/parser_test.go]: Makes sense, the part about thinking up names gets silly. Maybe either the fact that `t.Run` isn't needed or printing the output in `t.Error` can improve things. I really like having a self-contained function and I'm trying to converge on a format where I can have a template/boiler that works for any test. Spending cognitive effort on separating inputs into different blocks or deciding on whether the error format should be different for a test case, etc. all makes testing more painful (which happens in that linked example, but not directly related to the point we're discussing here). I _think_ having the testing always print out input in `t.Error` would satisfy all the things here.

Comment (unknwon) [reply internal/search/parser_test.go]: I think the purpose of name is actually a description, of the intention for the case. Otherwise, we won't be sure what we're trying to test in the future.

PS: If you just give an empty string, go test numbers it automatically.
